src/jit/flowgraph.cpp

```diff
@@ -22472,3 +22476,652 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+        assert(lastBlock != nullptr);
+        BasicBlock* const prevBlock = firstBlock->bbPrev;
+        BasicBlock*       nextBlock = lastBlock->bbNext;
+        assert(prevBlock != nullptr);
+        unsigned regionBBCount   = 0;
+        unsigned regionStmtCount = 0;
+        bool     hasFinallyRet   = false;
+        bool     isAllRare       = true;
+        bool     hasSwitch       = false;
+
+        for (const BasicBlock* block = firstBlock; block != nextBlock; block = block->bbNext)
+        {
+            if (block->bbJumpKind == BBJ_SWITCH)
+            {
+                hasSwitch = true;
+                break;
+            }
+
+            regionBBCount++;
+
+            // Should we compute statement cost here, or is it
+            // premature...? For now just count statements I guess.
+            for (GenTreeStmt* stmt = block->firstStmt(); stmt != nullptr; stmt = stmt->gtNextStmt)
+            {
+                regionStmtCount++;
+            }
+
+            hasFinallyRet |= (block->bbJumpKind == BBJ_EHFINALLYRET);
+            isAllRare &= block->isRunRarely();
+        }
+
+        // Skip cloning if the finally has a switch.
+        if (hasSwitch)
+        {
+            JITDUMP("Finally in EH region %u has a switch; skipping.\n", XTnum);
+            continue;
+        }
+
+        // Skip cloning if the finally must throw.
+        if (!hasFinallyRet)
+        {
+            JITDUMP("Finally in EH region %u does not return; skipping.\n", XTnum);
+            continue;
+        }
+
+        // Skip cloning if the finally is rarely run code.
+        if (isAllRare)
+        {
+            JITDUMP("Finally in EH region %u is run rarely; skipping.\n", XTnum);
+            continue;
+        }
+
+        JITDUMP("Try-finally EH region %u is a candidate for finally cloning: \n"
+                "%u blocks, %u statements\n",
+                XTnum, regionBBCount, regionStmtCount);
+
+        // Find all the call finallys that invoke this finally. Note
+        // some of them may only be reachable via an exceptional path.
+        BasicBlock* firstCallFinallyBlock = nullptr;
+        BasicBlock* lastCallFinallyBlock  = nullptr;
+        ehGetCallFinallyBlockRange(XTnum, &firstCallFinallyBlock, &lastCallFinallyBlock);
+
+        // Keep track of the blocks where the normally invoked
+        // (non-exceptional) finally is called from and returns to.
+        // Determine where to put the cloned finally, and what
+        // protected region (if any) it belongs to. Note because of
+        // the screening above, we know the finally is not inside any
+        // other handler region.
+        //
+        // We assume the lexically first callfinally is the one that
+        // corresponds to the normal try exit point.
+        BasicBlock* normalCallFinallyBlock  = nullptr;
+        BasicBlock* normalCallFinallyReturn = nullptr;
+        BasicBlock* cloneInsertAfter        = lastCallFinallyBlock;
+
```

Comment:
My understanding was that when there were no call finally thunks, the callfinally and paired return would be within the try region, and (presuming we successfully identified the normal exit pair) would be right at the end of the region.

Easy enough to change this to the try region end, if it seems clearer that way.

First block of the clone always uses `fgNewBBinRegion`  with the enclosing try region, no matter what we use for a hint block. The main concern with the hint is to try and avoid jumping in and out of the cloned finally; if it's placed poorly nothing later seems to be able to fix it.